File: JAVA/src/_Array/_ArrayQues/Searching/BinarySearchQues/LastOccurence.java
package _Array._ArrayQues.Searching.BinarySearchQues;

public class LastOccurence {
    public  static int LastOccurence(int size, int [] arr,int key)
    {
        int mid, ans =-1, s=0,e=size-1;

        while(s<=e)
        {
            mid=(s+e)/2;
            if (arr[mid]==key)
            {
                ans = mid;
                s=mid+1;
            }
            else if(arr[mid]<key) s=mid+1;
            else if (arr[mid]>key) e=mid-1;
        }
        return ans;
    }
    public  static int firstOccurence(int size, int [] arr,int key)
    {
        int mid, ans =-1, s=0,e=size-1;

        while(s<=e)
        {
            mid=(s+e)/2;
            if (arr[mid]==key)
            {
                ans = mid;
                e=mid-1;
            }
            else if(arr[mid]<key) s=mid+1;
            else if (arr[mid]>key) e=mid-1;
        }
        return ans;
    }

    public static int CountElements(int size, int arr[], int key)
    {
        int first = firstOccurence(size,arr,key);
        int Last= LastOccurence(size,arr,key);
        if(first==-1 || Last==-1) return -1;
        return Last-first+1;
    }
    public static void  main(String [] args)
    {
        int arr[] = {1,2,3,3,3,3,4,4,6,7};
        int size=9;
        System.out.println("Count Element : " + CountElements(size,arr,4));

    }

}

File: CPP/Array/lastOccurence.cpp
#include <iostream>
using namespace std;

int firstOccurence(int arr[], int size, int key)
{

    int s = 0, h = size - 1;
    while (s < h)
    {
        int mid = (s + h) / 2;
        if (arr[mid] == key && arr[mid - 1] > key)
        {
            return mid;
        }
        else if (arr[mid] == key && arr[mid - 1] == key)
        {
            s = 0;
            h = mid - 1;
        }
        else if (arr[mid] > key)
        {
            h = mid - 1;
        }
        else
        {
            s = mid + 1;
        }
    }
}
int lastOccurence(int arr[], int size, int key)
{

    int s = 0, h = size - 1;
    while (s < h)
    {
        int mid = (s + h) / 2;
        if (arr[mid] == key && arr[mid + 1] > key)
        {
            return mid;
        }
        else if (arr[mid] == key && arr[mid + 1] == key)
        {
            s = mid + 1;
        }
        else if (arr[mid] > key)
        {
            h = mid - 1;
        }
        else
        {
            s = mid + 1;
        }
    }
}
int main()
{
    int arr[7] = {1, 2, 3, 3, 3, 3, 4};
    int first = firstOccurence(arr, 7, 3);
    int last = lastOccurence(arr, 7, 3);
    int ans = last - first + 1;
    cout << ans;
}